src/OpenTelemetry/Logs/OpenTelemetryLogger.cs

```diff
@@ -77,22 +90,30 @@ public bool IsEnabled(LogLevel logLevel)
 
         public IDisposable BeginScope<TState>(TState state) => this.ScopeProvider?.Push(state) ?? NullScope.Instance;
 
-        private IReadOnlyList<KeyValuePair<string, object?>> ParseState<TState>(TState state)
+        private static IReadOnlyList<KeyValuePair<string, object?>> ParseState<TState>(LogRecord logRecord, TState state)
         {
             if (state is IReadOnlyList<KeyValuePair<string, object?>> stateList)
             {
                 return stateList;
             }
             else if (state is IEnumerable<KeyValuePair<string, object?>> stateValues)
             {
-                return new List<KeyValuePair<string, object?>>(stateValues);
+                var attributeStorage = logRecord.AttributeStorage;
+                if (attributeStorage == null)
+                {
+                    return logRecord.AttributeStorage = new List<KeyValuePair<string, object?>>(stateValues);
+                }
+                else
+                {
+                    attributeStorage.AddRange(stateValues);
+                    return attributeStorage;
+                }
             }
             else
             {
-                return new List<KeyValuePair<string, object?>>
-                {
-                    new KeyValuePair<string, object?>(string.Empty, state),
-                };
+                var attributeStorage = logRecord.AttributeStorage ??= new List<KeyValuePair<string, object?>>(8);
```

Comment:
Why 8?

Reply:
Yeah, I guess we could stay consistent with scopes. We could use the constant variable here and have it updated to 8 if need be.